File: src/main/java/com/cardlay/techinterview/cardmanagement/entity/Card.java
package com.cardlay.techinterview.cardmanagement.entity;

import javax.persistence.*;
import java.math.BigDecimal;
import java.util.Set;

@Entity
@Table(name = "cards")
public class Card {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String cardNumber;
    private BigDecimal balance;

    @ManyToOne
    @JoinColumn(name = "card_holder_id", nullable = false)
    private CardHolder cardHolder;

    @OneToMany(mappedBy = "card")
    private Set<Transaction> transactions;

    public void setId(Long id) {
        this.id = id;
    }

    public Long getId() {
        return id;
    }

    public String getCardNumber() {
        return cardNumber;
    }

    public void setCardNumber(String cardNumber) {
        this.cardNumber = cardNumber;
    }

    public BigDecimal getBalance() {
        return balance;
    }

    public void setBalance(BigDecimal balance) {
        this.balance = balance;
    }

    public CardHolder getCardHolder() {
        return cardHolder;
    }

    public void setCardHolder(CardHolder cardHolder) {
        this.cardHolder = cardHolder;
    }

    public Set<Transaction> getTransactions() {
        return transactions;
    }

    public void setTransactions(Set<Transaction> transactions) {
        this.transactions = transactions;
    }

    public void addTransaction(Transaction transaction) {
        if (this.transactions.contains(transaction)) {
            return;
        }

        transaction.setCard(this);
        this.transactions.add(transaction);
    }
}

File: src/main/kotlin/com/cardlay/techinterview/cardmanagement/entity/Card.kt
package com.cardlay.techinterview.cardmanagement.entity

import java.math.BigDecimal
import javax.persistence.*

@Entity
@Table(name = "cards")
class Card {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0
    var cardNumber: String = ""
    var balance: BigDecimal = BigDecimal.ZERO

    @ManyToOne
    @JoinColumn(name = "card_holder_id", nullable = false)
    var cardHolder: CardHolder = CardHolder()

    @OneToMany(mappedBy = "card")
    private var transactions: MutableSet<Transaction> = mutableSetOf()

    fun getTransactions(): Set<Transaction> {
        return transactions
    }

    fun setTransactions(transactions: MutableSet<Transaction>) {
        this.transactions = transactions
    }

    fun addTransaction(transaction: Transaction) {
        if (transactions.contains(transaction)) {
            return
        }
        transaction.card = this
        transactions.add(transaction)
    }
}